libandroid-navigation/src/main/java/com/mapbox/navigation/DefaultTripSession.kt

```diff
@@ -0,0 +1,133 @@
+package com.mapbox.navigation
+
+import android.location.Location
+import android.os.Handler
+import android.os.Looper
+import com.mapbox.android.core.location.LocationEngine
+import com.mapbox.android.core.location.LocationEngineCallback
+import com.mapbox.android.core.location.LocationEngineRequest
+import com.mapbox.android.core.location.LocationEngineResult
+import com.mapbox.navigation.base.route.Route
+import com.mapbox.navigation.base.trip.RouteProgress
+import com.mapbox.navigation.base.trip.TripService
+import com.mapbox.navigation.base.trip.TripSession
+import com.mapbox.navigation.navigator.MapboxNativeNavigator
+import java.lang.ref.WeakReference
+import java.util.Date
+import java.util.concurrent.CopyOnWriteArrayList
+
+class DefaultTripSession(
+    override val tripService: TripService,
+    override val locationEngine: LocationEngine,
+    override val locationEngineRequest: LocationEngineRequest,
+    private val navigator: MapboxNativeNavigator,
+    private val mainHandler: Handler,
+    private val workerHandler: Handler
+) : TripSession {
+
+    override var route: Route? = null
+        set(value) {
+            field = value
+            if (value != null) {
+                workerHandler.post { navigator.setRoute(value) }
+            }
+        }
+
+    private val mainLocationCallback = MainLocationCallback(this)
+
+    private val locationObservers = CopyOnWriteArrayList<TripSession.LocationObserver>()
+    private val routeProgressObservers = CopyOnWriteArrayList<TripSession.RouteProgressObserver>()
+
+    var rawLocation: Location? = null
```

Comment:
i'm trying to understand how end user would get access to `rawLocation` and `enhancedLocation` if the trip is referenced via `TripSession`? should they be part of the contract?

Reply:
Very good question. These fields cannot be part of the contract, because that would either make them immutable or expose a public setter.

I still wanted to expose a convenient method to get the lastest location/progress without registering the observer, but you're right, it should be part of a contract but in a form of a getter method, not a field.